api/app.py

```diff
@@ -17,72 +18,168 @@
+    return jsonify({'message': f'Recipe with ID {recipe_id} deleted successfully'}), 204
+
 
 @app.route('/api/recipes/<int:recipe_id>', methods=['PUT'])
 def edit_recipe(recipe_id):
-    for recipe in recipes:
-        if recipe_id == recipe['id']:
-            recipe['name'] = categories if (categories := request.json.get('name')) else recipe['name']
-            recipe['duration'] = categories if (categories := request.json.get('duration')) else recipe['duration']
-            recipe['pictures'] = categories if (categories := request.json.get('pictures')) else recipe['pictures']
-            recipe['instructions'] = categories if (categories := request.json.get('instructions')) else \
-                recipe['instructions']
-            recipe['categories'] = categories if (categories := request.json.get('categories')) else \
-                recipe['categories']
-            recipe['ingredients'] = categories if (categories := request.json.get('ingredients')) else \
-                recipe['ingredients']
-            return recipe
-    return jsonify({"error": "404 not found"}), 404
+    recipe = db.session.get(Recipe, recipe_id)
+    if not recipe:
+        return jsonify({'error': 'Recipe not found'}), 404
+
+    data = request.json
+
+    if data['name']:
+        recipe.name = data['name']
+    if data['duration']:
+        recipe.duration = data['duration']
+    if data['pictures']:
+        recipe.pictures = ",".join(data['pictures'])
+    if data['instructions']:
+        recipe.instructions = data['instructions']
+
+    if data['categories']:
+        recipe.categories.clear()
+
+        for cat_data in data['categories']:
+            category = Category.query.get(cat_data["name"])
+            if category:
+                recipe.categories.append(category)
+            else:
+                return jsonify({'error': f'category {cat_data["name"]} not found'}), 400
+
+    if data['ingredients']:
+        Ingredient.query.filter_by(recipe_id=recipe.id).delete()
+        for ing in data['ingredients']:
+            new_ing = Ingredient(
+                name=ing["name"],
+                quantity=ing["quantity"],
+                unit=ing.get("unit"),
+                recipe_id=recipe.id
+            )
+            db.session.add(new_ing)
+
+    db.session.commit()
+
+    return jsonify(recipe.as_dict()), 201
+
+
+@app.route('/api/categories/', methods=['GET'])
+def get_category():
+    categories = []
+
+    for category in db.session.query(Category).all():
+        categories.append(category.as_dict())
+
+    return jsonify(categories), 200
+
+
+@app.route('/api/categories/', methods=['POST'])
+def create_category():
+    data = request.json
+    category = Category.query.filter_by(name=data['name']).first()
+    if not category:
+        category = Category(
+            name=data['name'],
+            color=data.get('color', '#848482'))
+        db.session.add(category)
+    else:
+        return "Category already exists"
```

Comment:
We should always return data in the same format, so the error message should be in json format and also return a status code:
```python
return jsonify({'error': f'category already exists'}), 400
```